### PR TITLE
Refactor resource classes

### PR DESCRIPTION
**Pull request by whyscream**
We decided to overhaul the API interface of the client resources to make it more logical. No more methods on the client that don't work, f.i `client.chargebacks.get(chargeback_id)`. 

See [the refactoring docs](https://github.com/mollie/mollie-api-python/blob/255-refactor-resources/v3-api-changes.md) for details.

Related to #255

### CODE REVIEW

File: tests/test_orders.py [hunk 1]
```diff
@@ -83,6 +83,12 @@ def test_get_order(client, response):
         order.payments
 
 
+def test_get_order_invalid_id(client):
+    with pytest.raises(IdentifierError) as excinfo:
+        client.orders.get("invalid")
+    assert str(excinfo.value) == "Invalid order ID 'invalid', it should start with 'ord_'."
```
Comment (geertjanvdenbosch) [inline tests/test_orders.py:86]: What happens if the order ID starts with `ord_` but the rest of the ID is invalid?

**Review by geertjanvdenbosch — COMMENTED** [review]
Lines 7 to 9 of the class Capture (objects/capture.py) are not covered anymore by the unit tests. They were before, why is this changed?

File: tests/test_payments.py [hunk 2]
```diff
@@ -59,10 +59,20 @@ def test_cancel_payment(client, response):
     assert canceled_payment.id == PAYMENT_ID
 
 
-def test_cancel_payment_invalid_id(client):
+def test_payment_invalid_id(client):
```
Comment (geertjanvdenbosch) [inline tests/test_payments.py:62]: The other resources contain basically the same tests (checking if `get` or `update` (or any other method) throws an IdentifierError). However, they all do so in separate methods. Could you place the tests for `get` and `update` in separate methods?

File: mollie/api/resources/base.py [hunk 3]
```diff
@@ -36,37 +49,34 @@ def perform_api_call(self, http_method, path, data=None, params=None):
...
+        if not message:
+            message = f"Invalid {name} '{resource_id}', it should start with '{self.RESOURCE_ID_PREFIX}'."
 
         if not resource_id or not str(resource_id).startswith(self.RESOURCE_ID_PREFIX):
             raise IdentifierError(message)
 
     @staticmethod
     def extract_embed(params):
         """Extract and parse the embed parameter from the request."""
+        # TODO Remove this
```
Comment (geertjanvdenbosch) [inline mollie/api/resources/base.py]: Is this TODO intentional?

File: mollie/api/resources/base.py [hunk 4]
```diff
@@ -36,37 +49,34 @@ def perform_api_call(self, http_method, path, data=None, params=None):
                 )
         return result
 
-    def _validate_resource_id(self, resource_id: str, name: str = "", message: str = ""):
-        resource_name = name or "Identifier"
-        message = (
-            message or f"Invalid {resource_name} '{resource_id}', it should start with '{self.RESOURCE_ID_PREFIX}'."
-        )
+    def validate_resource_id(self, resource_id: str, name: str = "Identifier", message: str = ""):
```
Comment (geertjanvdenbosch) [inline mollie/api/resources/base.py]: Why did you remove the underscore from the method name? It still seems to be a private method even though it is inherited by the other resources.

File: mollie/api/resources/chargebacks.py [hunk 5]
```diff
@@ -1,19 +1,28 @@
...
-    def get(self, chargeback_id, **params):
-        """Verify the chargeback ID and retrieve the chargeback from the API."""
-        if not chargeback_id or not chargeback_id.startswith(self.RESOURCE_ID_PREFIX):
-            raise IdentifierError(
-                f"Invalid chargeback ID: '{chargeback_id}'. "
-                f"A chargeback ID should start with '{self.RESOURCE_ID_PREFIX}'."
-            )
+
+class Chargebacks(ChargebacksBase, ResourceListMixin):
+    pass
```
Comment (geertjanvdenbosch) [inline mollie/api/resources/chargebacks.py:20]: For the class Refunds (resource/refunds.py) you've added a docstring. Could you do the same for this class?

File: mollie/api/client.py [hunk 6]
```diff
@@ -295,6 +265,10 @@ def setup_oauth_authorization_response(self, authorization_response):
         self.set_token(self.access_token)
         return self.access_token
 
+    # TODO Implement https://docs.mollie.com/reference/oauth2/revoke-token
```
Comment (geertjanvdenbosch) [inline mollie/api/client.py:268]: Is this TODO intentional?

File: mollie/api/objects/organization.py [hunk 7]
```diff
@@ -43,3 +43,7 @@ def vat_regulation(self):
     @property
     def dashboard(self):
         return self._get_link("dashboard")
+
+    # TODO: Implement https://docs.mollie.com/reference/v2/organizations-api/get-partner
```
Comment (geertjanvdenbosch) [inline mollie/api/objects/organization.py:47]: Is this TODO intentional?

File: mollie/api/resources/order_lines.py [hunk 8]
```diff
@@ -1,48 +1,110 @@
...
+
+__all__ = [
+    "OrderLines",
+]
 
 
 class OrderLines(ResourceBase):
-    order_id = None
+    """
+    OrderLine resource class
```
Comment (geertjanvdenbosch) [inline mollie/api/resources/order_lines.py:15]: This should be plural since we're talking about the OrderLines class.

**Review by geertjanvdenbosch — CHANGES_REQUESTED** [review]
I've seen quite a few `TODO` comments in the PR. Is that something we wish to have on master or should that be adressed first?

**Comment by whyscream** [comment]
> Lines 7 to 9 of the class Capture (objects/capture.py) are not covered anymore by the unit tests. They were before, why is this changed?

The `get_resource_class()` is only used in an `ObjectList`: https://github.com/mollie/mollie-api-python/blob/255-refactor-resources/mollie/api/objects/list.py#L87-L99. It seems that we don't have tests for `get_next()` and `get_previous()` for Captures. We have proper testing for list handling though. Maybe we can add a generic test later that tests listing behavior for all objects. Doesn;t seem critital to me.

File: mollie/api/objects/subscription.py [hunk 9]
```diff
@@ -107,6 +110,19 @@ def customer(self):
         url = self._get_link("customer")
         return self.client.customers.from_url(url)
 
+    @property
+    def customer_id(self):
+        """
+        Retrieve the customer id from the customer link.
+
+        The customer_id is not available as a direct subscription property,
+        but we need it toimplement various features. The only option is to extract it from the link.
```
[the remark this answers is not in the document]
Comment (whyscream) [reply mollie/api/objects/subscription.py:119]: Fixed.

File: tests/test_customer_subscriptions.py [hunk 10]
```diff
@@ -138,41 +153,55 @@ def test_create_customer_subscription(client, response):
...
-    subscription = client.customer_subscriptions.with_parent_id(CUSTOMER_ID).create(data=data)
+    customer = client.customers.get(CUSTOMER_ID)
+    subscription = customer.subscriptions.create(data=data)
     assert isinstance(subscription, Subscription)
     assert subscription.id == SUBSCRIPTION_ID
 
 
 def test_update_customer_subscription(client, response):
     """Update existing subscription of a customer."""
+    response.get(f"https://api.mollie.com/v2/customers/{CUSTOMER_ID}", "customer_single")
```
[the remark this answers is not in the document]
Comment (whyscream) [reply tests/test_customer_subscriptions.py:164]: No stretch. I have several ideas on improving this in general: incorporating the base part of the URL in the `response` fixture, and not using the production URL (but `https://api.mollie.test/` or something like that). It seems not in scope for this PR, so I'd opt for doing that later.

File: tests/test_orders.py (hunk 1 above)
Comment (whyscream) [reply tests/test_orders.py:86]: The ID is just an identifier, invalid here seems to mean that you send `ord_invalid` or `ord_56789` in stead of `ord_12345`?

Then it;ll be sent to the API, and that will probably return a HTTP 404 error (depending which API operation you're doing). The client will then use that to raise a `RequestError`.

File: tests/test_payments.py (hunk 2 above)
Comment (whyscream) [reply tests/test_payments.py:62]: Fixed.

File: mollie/api/resources/base.py (hunk 3 above)
Comment (whyscream) [reply mollie/api/resources/base.py]: Yes, all the embed stuff is removed in commit c0803b693b93554a92664fddc0e87326018f248b

File: mollie/api/resources/base.py (hunk 4 above)
Comment (whyscream) [reply mollie/api/resources/base.py]: I thought this could be useful for developers implementing the library to validate user input and such.

File: mollie/api/client.py (hunk 6 above)
Comment (whyscream) [reply mollie/api/client.py:268]: Yes, but the change is not in scope for this PR (or a blocker for an upcoming release).

File: mollie/api/objects/organization.py (hunk 7 above)
Comment (whyscream) [reply mollie/api/objects/organization.py:47]: Yes, but the change is not in scope for this PR (or a blocker for an upcoming release).

File: mollie/api/resources/permissions.py [hunk 11]
```diff
@@ -2,18 +2,23 @@
...
-    def get_resource_object(self, result):
+
+class Permissions(ResourceGetMixin, ResourceListMixin):
+    def get_resource_object(self, result: dict) -> Permission:
         return Permission(result, self.client)
 
     @staticmethod
-    def validate_permission_id(permission_id):
+    def validate_permission_id(permission_id: str):
+        # TODO Maybe we should just include a list of supported permissions?
```
[the remark this answers is not in the document]
Comment (whyscream) [reply mollie/api/resources/permissions.py:20]: Yes, but the change is not in scope for this PR (or a blocker for an upcoming release). We can improve the validation later, if we want to.

**Comment by whyscream** [comment]
> I've seen quite a few `TODO` comments in the PR. Is that something we wish to have on master or should that be adressed first?

Most TODOs are there to resolve in a later release. They're new features, missing functionality or nice-to-haves, but not part of this refactor or a blocker for a release. We can add github issues for all TODO items after the release?

File: tests/test_orders.py (hunk 1 above)
Comment (geertjanvdenbosch) [reply tests/test_orders.py:86]: Ok, gotcha.

**Review by geertjanvdenbosch — APPROVED** [review]
